Support distributing by multiple columns

JelteF opened the pull request:
DESCRIPTION: Support distributing by multiple columns

IMPORTANT: Not finished at all yet

This has been initial work to support having multiple distribution columns. I'm opening this PR so that it's not lost and once we prioritize this we can continue on top of this. 

What this currently implements is roughly the following:
1. Speechifying multiple distribution columns in citus_create_distributed_table, which are saved in the catalog tables.
2. Checking if tables are colocated based when they have multiple distribution columns

What's notably not implemented:
1. Router planner checking for both columns being there
2. Actually distributing by 2 columns, for now this still only distributes by the first column. The second column is currently just an unnecessary planning restriction.
3. All of the functions that are not checked off on the checklist in the new file called `multi_column.txt`

JelteF committed:
Support distributing by multiple columns

JelteF committed:
EnsureRelationCanBeDistributed

JelteF committed:
Add list of functions that still need to be modified

JelteF committed:
BuildCitusTableCacheEntry

JelteF committed:
DistributionCreateCommand

JelteF committed:
Use empty partition column list instead of partition column list with NULL

JelteF committed:
Only read from pg_dist_partkeys column

JelteF committed:
Implement DistPartitionKeys

JelteF committed:
Refactor TargetListOnPartitionColumn

JelteF committed:
TargetListOnPartitionColumns

JelteF committed:
RestrictionEquivalenceForPartitionKeys

JelteF committed:
PartitionColumns

JelteF committed:
LocalJoin

hslightdb commented:
what's caused not continue?

hslightdb commented:
we have the plan to Support distributing by multiple columns, any suggestion?

JelteF commented:
@hslightdb mostly other features taking priority. One way to work around it is to distribute by a single column that contains a composite type. This type must implement a few functions like, hash, equal etc.

hslightdb commented:
> @hslightdb mostly other features taking priority. One way to work around it is to distribute by a single column that contains a composite type. This type must implement a few functions like, hash, equal etc.

now, we plan to support distributing by multiple columns, The main idea is to store “nodeToString(col1)+";"+nodeToString(col2)+...” in pg_dist_partition, then split the string when getting it, and generate a list of var structure (DistPartitionKey/PartitionColumn), and then modify the function involved to process the list of var(like LocalJoin, JoinOrderForTable,... ) and list of distributionKey  index. The hash function uses the hash function of the first field in the distribution key.  


Is the idea above similar to yours?  

Whether all the functions that need to be modified are included in multi_column.txt, we found that some functions that need to be modified by our idea are not included in multi_column.txt.

hslightdb commented:
> 



> > @hslightdb mostly other features taking priority. One way to work around it is to distribute by a single column that contains a composite type. This type must implement a few functions like, hash, equal etc.
> 
> now, we plan to support distributing by multiple columns, The main idea is to store “nodeToString(col1)+";"+nodeToString(col2)+...” in pg_dist_partition, then split the string when getting it, and generate a list of var structure (DistPartitionKey/PartitionColumn), and then modify the function involved to process the list of var(like LocalJoin, JoinOrderForTable,... ) and list of distributionKey index. The hash function uses the hash function of the field in the distribution key and then combine it by hash_combine.
> 
> Is the idea above similar to yours?
> 
> Whether all the functions that need to be modified are included in multi_column.txt, we found that some functions that need to be modified by our idea are not included in multi_column.txt.

@JelteF  Looking at your revision, our thinking is similar to yours. If it is as I understand it, then our revision can be consistent with this PR. Please give us a reply when you have time.

JelteF commented:
What you're discribing is similar, but in this draft PR I instead added a new column of type `[]string`. I think that's cleaner than using a `;` as separator in a string.

multi_column.txt is an incomplete list, I added to it whenever I found a function that needed changing and added them to this list so I did not forget. 

Please create a PR if you get it working (or close to working).

linlom025 commented:
I want to create a specific sharding selection strategy for Citus. However, since Citus currently doesn’t support multiple sharding keys, I'm unsure how to proceed. If anyone could provide ideas or solutions, I would be very grateful.
